third_party/binutils/test_decenc/drdecode_decenc_x86.c

```diff
@@ -0,0 +1,3786 @@
+void
+test_x86_64_mpx_add_bnd_prefix_s_asm(void);
+void
+test_x86_64_mpx_addr32_s_asm(void);
+void
+test_x86_64_mpx_branch_1_s_asm(void);
+void
+test_x86_64_mpx_branch_2_s_asm(void);
+void
+test_x86_64_mpx_s_asm(void);
+void
+test_x86_64_nops_s_asm(void);
+void
+test_x86_64_notrack_s_asm(void);
+void
+test_x86_64_opcode_s_asm(void);
+void
+test_x86_64_optimize_1_s_asm(void);
+void
+test_x86_64_optimize_2_s_asm(void);
+void
+test_x86_64_optimize_3_s_asm(void);
+void
+test_x86_64_optimize_4_s_asm(void);
+void
+test_x86_64_opts_s_asm(void);
+void
+test_x86_64_pcrel_s_asm(void);
+void
+test_x86_64_pr20141_s_asm(void);
+void
+test_x86_64_prefetchwt1_s_asm(void);
+void
+test_x86_64_pseudos_s_asm(void);
+void
+test_x86_64_ptwrite_s_asm(void);
+void
+test_x86_64_rdpid_s_asm(void);
+void
+test_x86_64_rdrnd_s_asm(void);
+void
+test_x86_64_rdseed_s_asm(void);
+void
+test_x86_64_reg_s_asm(void);
+void
+test_x86_64_relax_1_s_asm(void);
+void
+test_x86_64_rep_suffix_s_asm(void);
+void
+test_x86_64_rep_s_asm(void);
+void
+test_x86_64_rip_s_asm(void);
+void
+test_x86_64_rtm_s_asm(void);
+void
+test_x86_64_segment_s_asm(void);
+void
+test_x86_64_segovr_s_asm(void);
+void
+test_x86_64_sha_s_asm(void);
+void
+test_x86_64_sib_s_asm(void);
+void
+test_x86_64_simd_s_asm(void);
+void
+test_x86_64_size_5_s_asm(void);
+void
+test_x86_64_sse_noavx_s_asm(void);
+void
+test_x86_64_sse2avx_s_asm(void);
+void
+test_x86_64_sse3_s_asm(void);
+void
+test_x86_64_sse4_1_s_asm(void);
+void
+test_x86_64_sse4_2_s_asm(void);
+void
+test_x86_64_ssse3_s_asm(void);
+void
+test_x86_64_stack_s_asm(void);
+void
+test_x86_64_suffix_s_asm(void);
+void
+test_x86_64_tbm_s_asm(void);
+void
+test_x86_64_unwind_s_asm(void);
+void
+test_x86_64_vaes_s_asm(void);
+void
+test_x86_64_vex_lig_2_s_asm(void);
+void
+test_x86_64_vgather_check_s_asm(void);
+void
+test_x86_64_vpclmulqdq_s_asm(void);
+void
+test_x86_64_waitpkg_s_asm(void);
+void
+test_x86_64_wbnoinvd_s_asm(void);
+void
+test_x86_64_xop_s_asm(void);
+void
+test_x86_64_xsave_s_asm(void);
+void
+test_x86_64_xsavec_s_asm(void);
+void
+test_x86_64_s_asm(void);
+void
+test_x86_64_reloc64_s_asm(void);
+void
+test_xmmhi32_s_asm(void);
+void
+test_xmmhi64_s_asm(void);
+void
+test_xop_s_asm(void);
+void
+test_xop32reg_s_asm(void);
+void
+test_xsave_s_asm(void);
+void
+test_xsavec_s_asm(void);
+
+/* Arbitrary pc for pc-relative operands for consistent output */
+#    define ORIG_PC ((app_pc)0x10000000)
+
+static void
+test_s(byte *subtest_asm)
+{
+    byte *orig_pc;
+    instr_t *instr = instr_create(GD);
+    byte buf[16];
+    byte *dec_pc = (byte *)subtest_asm;
+    byte *start = dec_pc;
+    int count = 0;
+    disassemble_set_syntax(DR_DISASM_ATT);
+    while (*(uint64 *)dec_pc != 0x0102030405060708) {
+        instr_reset(GD, instr);
+        orig_pc = dec_pc;
+        decode_from_copy(GD, dec_pc, ORIG_PC + (dec_pc - start), instr);
+        dec_pc = disassemble_from_copy(GD, dec_pc, ORIG_PC + (dec_pc - start), STDOUT,
```

Comment:
The order makes this look like inefficient code: if you've already decoded the instr you'd use instr_disassemble() instead of re-decoding it by using a raw disassemble.  I would swap and do the disassemble first: the decode belongs with the two lines below.

Reply:
Done.